### ISSUE TITLE
Should Array.idxType return a tuple for multi-dimensional arrays?

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#19141]
This was brought up during the Array Module re-review (https://github.com/Cray/chapel-private/issues/2955).

Right now if you have (say) a 3D array of ints and query array.idxType you'll just get back `int` but maybe it would be better if returned a tuple like: `3*int`?

This would allow doing things like the following:

```
var i : A.idxType;
i = A.domain.first;
```

Some other thoughts\questions:
* In the 1D case should this function return the value or a 1-tuple of the value?
* If we go down this route should we also have a function that returns the base type?
* Maybe we'd want the same behavior on `intIdxType` (although maybe we'll get rid of that method altogether?)

**Comment by bradcray** [comment]
I'm generally in favor of this proposal, where I'm imagining we'll:
* have something internally like `perDimIdxType` for rectangular domains/arrays to play the role that `idxType` does today
* continue to have the `domain` type take this `perDimIdxType` in its initializer, so that domain signatures like the ones we use today will still work

> In the 1D case should this function return the value or a 1-tuple of the value?
This has been a longstanding pain-point, and we've tried (and failed) to support coercions between `1*t` and `t` to make it less painful, but I think that for consistency with where we are today, it would need to simply be the value, not a 1-tuple of the value.  I.e., in `forall i in My1DArray.domain ...`, `i` is of type `int`, not `1*int`

> If we go down this route should we also have a function that returns the base type?

Whether or not we have a method for this, if we continue to store that type as a field within the record, the user could presumably access it that way (unless we take pains to hide it from them).

> Maybe we'd want the same behavior on intIdxType (although maybe we'll get rid of that method altogether?)

This would actually be a good exercise for my TODO to see whether anything outside of the internal code needs access to this, or whether it's internal only (and generally, this issue is one I'd be very willing to help with, in my copious development time).

**Comment by vasslitvinov** [comment]
We should apply the solution, whichever it is, to domains as well.

**Comment by stonea** [comment]
Although the type constructor for arrays is not user exposed (array declarations de-sugar to it and it gets used explicitly in module code) this is not true for domains.

In the case of a domain think of something like

```chpl
var D: domain(rank=3, idxType=int);
```

If we have `.idxType` (as a query on arrays/domains) return a tuple then there's an asymmetry in how it's used there and in the type constructor.

We could imagine requiring idxType to be passed in as a tuple like:

```chpl
var D: domain(rank=3, idxType=3*int);
```

But there's sort of an ugly redundancy there.

We could also imagine removing the rank parameter altogether and expressing a 3-dimensional domain like this:

```chpl
var D: domain(idxType=3*int);
```

But this is ambiguous with an associative domain of `3*int`.

**Comment by bradcray** [comment]
The few thoughts I've had on this topic since coming up with the observation Andy summarized just above:

* we could continue to think of `idxType` as the per-dimension, lower-level index type for ranges/domains/arrays and introduce a different/friendlier way of getting the `3*int` answer that's wished for here.
  * we already have `index(D)` as a way of getting this today (also intended to semantically assert "And this is also a guaranteed index contained within D", but that was never implemented due to competing priorities and a few challenges... more on this below)
  * we could introduce a method with a different/friendlier name, like `indexType` or `getIndexType` or `getFullIndexType` or ...?  (Engin suggested `.index` but that felt too "noun-y" to me, like it would return an arbitrary index from A rather than its type)

  an obvious advantage to this approach is that it should not require changing nearly as much existing code since the role of `idxType` as type initializer argument and query wouldn't change

* I'll also state for the record that I don't find the middle of the three approaches above unworkable.  One slight benefit of it is that it would open a path towards potentially / maybe someday permitting different index types in different dimensions like `(int, int(8), uint(16))`.  That said, though we've kicked around the notion that we _could_ support this since the inception of the project, I'm not sure how naturally useful/desireable it is (relative to creating an array of arrays)—specifically, I can't recall any user wanting it.  And enabling that would lead to some implementation/representation challenges (e.g., we could no longer store an array of offsets, bounds, etc. within the array record since they're all different types... I guess we'd store those values as tuples with similar type as well, which might be fine).  So I guess I'm lukewarm on it at best.  And until we extended support to non-homogeneous tuple types as indices, it would just be a place for a user to make an error / do more typing (e.g., `domain(3, int)` would become `domain(3, 3*int)`).

Back on `index(D)`, Engin said something to the effect of "usually when we're querying properties of things, we do it as methods rather than as standalone functions."  At the time, I responded saying that I don't think of this as a standalone function, but more of a type specifier, like `range(int)` or `domain(2)` or `sparse subdomain(D)`.  The relationship to `subdomain` is actually pretty intentional where, in the early days, we imagined having a hierarchy of subdomains which would permit indices from a child to not have to be bounds-checked for the parent; and then a hierarchy of indices that worked the same way.  For example:

```chapel
var D = {1..n, 1..n};
var A: [D] real;
var D2 = D[2.., 2..];  // as a slice of D, so D2's type would be subdomain(D);
forall ind in D2 do  // ind's type is index(D2)
  A[ind] = 2.0;      // since ind is index(D2) and D2 is a subdomain(D), `ind` can be known to be in-bounds for `A`
```

so no bounds check is required, with or without `--fast`.  Similarly:

```chapel
var iD: index(D) = ...;
var iD2: index(D2) = ...;
iD = iD2;  // guaranteed safe since D2 is a subdomain(D)
iD2 = iD;  // requires a bounds check to make sure this is a safe assignment
```

As mentioned today, this is an idea that has largely been in limbo due to other priorities and one significant challenge:  In the presence of changing (particularly shrinking domains), what are the semantics for a variable of type `index(D)`.  For example:

```chapel
var D = {1..n, 1..n};
var A: [D] real;
var i: index(D) = (n, n);  // OK.  We can bounds-check `(n,n)` at assignment time, verify that it's within D, and know that it's safely an `index(D)`.
...A[i]...  // since i is an index(D), no bounds check is needed here
D = {1..n/2, 1..n/2};   // uh-oh...
...A[i]...  // now i is out of bounds, but its type is still index(D).
```

Specifically, we never came to a firm conclusion whether this should be:
* user error, not Chapel's fault, don't do that
* compiler must re-bounds-check all index(D) variables after D is resized (shrunk only, if/when it can determine that)...
  * proactively right after the resizing?
  * lazily, upon first use?
  * and what amount of state and/or compiler analysis would be required to implement either of these?
* or perhaps users just can't store `index(D)` variables for non-`const` domains D?  Or can't have them persist across re-assignments to D

One of my favorite examples of using `index(D)` was in a CSR-style implementation of sparse matrices where, in C, you just have a bunch of arrays of ints, each of which has a particular semantic meaning that's not particularly clear unless you've learned the pattern in the past.  Whereas with index types, you could imagine:

```chapel
var rows = 1..numRows;  // the rows the sparse matrix represents
var cols = 1..numCols;  // its columns
var bbox = {rows, cols};  // the conceptual bounds of the sparse matrix
var nnzDom = {1..nnz};  // the dense indices representing the nonzeroes
var rowStart: [rows] index(nnzDom);
var colIdx: [nnzDom] index(cols);
var elts: [nnzDom] eltType;
```

Then, when iterating from `rowStart[rows]..rowStart[rows+1]-1`, you could "know" that all indices were in-bounds when used to index into `colIdx` or `elts` since nnzDom is dense, both bounds are `index(nnzDom)` and all represented values are between those bounds.

Even without the bounds-checking semantics, using these type aliases results in code that is arguably more self-documenting which is why we've supported them as type aliases even without finishing that part of the story.  All that said, it's telling that a Chapel user as expert as Engin is not familiar with them (in part, possibly, because it's not often that a user declares explicitly typed index variables...  but arguably in part because we haven't shone a light on them, knowing that the bounds-checking aspects are largely unresolved).

**Comment by stonea** [comment]
To get a sense of what the fallout would be for this change I tried it out.  There's a branch for that showing the necessary changes here:

https://github.com/chapel-lang/chapel/compare/main...stonea:chapel:experiment_have_idxType_return_tuple_for_multidim

The PR passes paratests fine.  No real test updates were required beyond updating `.good` files for tests that are directly testing `idxType`.

Doing `grep -r "idxType" *` doesn't return anything when run on the main branch of the public repos for ChOp, ChplUltra's DistributedFFT, and Arkouda.  I also don't see it in the Champs code.

What's harder is doing the same with domain.  I spent a couple hours on it but wasn't able to determine every place that uses `domain.idxType` that I'd have to update to do something like `domain.scalarIdxType`.

So:
* It seems like changing the meaning of `idxType` for arrays can be done pretty easily
* It doesn't look like `.idxType` is commonly used in user code
* Updating `idxType` for domains should be possible to do but might require making a lot of painful modifications in our module code
* I can't imagine that we'd want to update one and not the other.

And finally:

Apart from any pain updating our code there might be other reasons why we wouldn't want to proceed with this change (like the fact that we use `idxType` as an argument name and that we already have `index(D)`).

**Comment by stonea** [comment]
To follow up from this after our module re-review meeting:

Firstly, we've decided that given the unsettled behavior for `index` as described in Brad's comment (https://github.com/chapel-lang/chapel/issues/19141#issuecomment-1165290801) we should mark it unstable for the time being.

As far as `idxType`, we think the right course of action here is to split it into two or three functions:

* Always returns a single type: like `Array.idxType` does today,
* It depends on rank: return a tuple of types when `rank > 1`,
* Always return a tuple

It may be that the right thing to do is have `Array.idxType` remain being the function that returns a single type but I think we'll want to see some proposed names for these 2-3 functions before settling on that.

Another option would be to have one (or more) of these functions folded together and use a param to dictate what kind of output you want.

Anyway, to get things started let me throw out some potential names that could be divvied up between these functions:

`idxType`, `baseIdxType`, `underlyingIdxType`, `rootIdxType`, `fullIdxType`, `idxVarType`, `ndIdxType`, `idxTypeTuple`

**Comment by vasslitvinov** [comment]
I like: `baseIdxType`, `tupleIdxType`, `idxType`, with idxType as the rank-dependent query.

My only reservation is that this would mean changing a lot of user code. OTOH... Arkouda does not reference idxType and maybe CHAMPS don't either, so this is not as much of an issue?

**Comment by vasslitvinov** [comment]
Now I like Engin's proposal. Indeed, we want something short and sweet for the initializer.

Alternatively we remove the `rank` argument of the initializer, so Engin's example now looks like `domain(2*uint)` or `domain(idxType=2*uint)`. For 1-d it could be either `idxType=uint` or `idxType=1*uint`. Hm, `idxType=uint` conflicts with associative domains so probably just the latter. Hmmmm the latter will also conflict with an associative domain over tuples. So maybe this alternative is not great.

If the alternative is not taken, we have the choice of telling users to use `tupleIdxType` or telling them to use `tuplify(fullIdxType)`. These two offer the same cognitive load. Although I prefer `tupleIdxType` as it is direct.

**Comment by bradcray** [comment]
I've been brainstorming on this since the meeting (before seeing the last few posts), and feel as though, if I were doing things from scratch, I'd probably do:

* `dimIdxType` — the per-dimension idxType
* `idxType` — the int / rank*int tuple solution
*  (and then wait on providing the always-tuple type until a motivating case came along

If I were trying to preserve `idxType` (and I'm currently split on how important I think that is), I'd either go with Engin's approach, or use `accessType` for the int/rank*int solution.

**Comment by vasslitvinov** [comment]
In today's deep dive, we concluded:
* `idxType` is the per-dimension type; this is appealing because we do not need to change the existing code base and because for now our multi-dimensional domains/arrays will always have the same idxType in each dimension
* `fullIdxType` is the type we get when iterating over a domain, i.e., scalar for 1-d and rank-tuple for multi-d; we favored this over `access[Idx]Type` because it needs to apply equally to arrays and domains and there is no "access" in domains
* re the type that is the rank-tuple of `idxType` -- hold off introducing a method to return that until there is a compelling use case

Elsewhere we recently decided to mark the index type `index(D)` as unstable due to its unclear status. It would not help in this case because we want to have the type of the domain indices that allows values that are outside the domain. This latter is needed for the `find` function in #21423.
